### ISSUE TITLE
add '~/bin' to search path for node & npm

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-ide-common#88]
If the IDEs are going to search in common `bin` directories for `node` and `npm`, I believe they should try `~/bin` *first*.

Advantages:

* Easy to set/override location of binaries without `sudo` and without writing to system directories (for me, just `ln -s ~/opt/node/bin/node ~/bin/node`)
* Easy to document
* Very easy to setup & verify compared to system properties and environment variables
* Can be done once, and will work for all IDEs and will not have to be re-configured each time a new version of Eclipse is installed

Disadvantages:

* None :) (If we're searching paths anyway, why not `~/bin`?)

For comparison, on macOS what I currently have is:

    launchctl setenv NODE_EXE ~/opt/node/bin/node

in `~/.bashrc`. This is not a persistent setting, so in the rare chance I reboot and launch Eclipse before opening a terminal, `node` doesn't work. Other options on macOS are much more complicated and difficult.

**Comment by jvasileff** [comment]
cc @davidfestal @quintesse

**Comment by quintesse** [comment]
Btw, in my idea of a perfect world this is how things would work when the IDE tries to execute something that could need `node` or `npm`:

 1. a method called `getNodeBinary()` (or `getNpmBinary()`) gets called
 2. check global IDE configuration for configured value of this path
 3. if the value exists check that the binary exists or if the value is empty search the `PATH` for the binary
 4. if the above fails open the dialog related to option 2 and let the user select the path to the binary
 5. return the path to the binary (or an exception if the user cancelled the dialog)

In that same perfect world the dialog mentioned in point 4 would have a "Find" button and it would be *that* button that goes looking through that hard-coded list of paths.

One could then argue if for the first time after installing the Ceylon plugin it would do that "Find" automatically without prompting the user or not.

But the idea of just always looking in a hard-coded list without being able to see what it is doing and where it found the binaries and no way to override that behaviour seems wrong to me.

**Comment by gavinking** [comment]
"Dialog"? C'mon, what is this, 1992?

**Comment by jvasileff** [comment]
> But the idea of just always looking in a hard-coded list without being able to see what it is doing and where it found the binaries and no way to override that behaviour seems wrong to me.

@quintesse I agree. But as a practical matter, any configuration technique that I have to redo for each IDE and each time I upgrade or re-download Eclipse really adds a lot of friction and IME creates a really bad user experience.

Setting the env var *is* something that has worked for *me* though.

**Comment by quintesse** [comment]
> "Dialog"? C'mon, what is this, 1992?

What? Then what do you call the preferences dialog if it's not a dialog?
I'm just saying that we can pop up the part of the IDE configuration where you can configure that option.

> any configuration technique that I have to redo for each IDE and each time I upgrade or re-download Eclipse really adds a lot of friction

Well that's why I said "One could then argue if for the first time after installing the Ceylon plugin it would do that "Find" automatically without prompting the user". You would then have the same experience we have now but more flexible in case the user wants something else or in case we really can't find the binaries.

**Comment by gavinking** [comment]
> What? Then what do you call the preferences dialog if it's not a dialog?

I do _not_ want to mess about with dialogs!

**Comment by quintesse** [comment]
We're using external binaries, they *need* to be configurable!

**Comment by gavinking** [comment]
> configurable

Configur_able_ is one thing. But we know there is an algorithm that works for 98% of users without requiring anything to be configured, and so we should use it by default. Anyway, David has already implemented that, so it's fine.
